crates/ere-risc0/src/prove/bento.rs

```diff
@@ -0,0 +1,230 @@
+    let rest_api_tag = format!("ere-risc0/rest_api:{SDK_VERSION}");
+
+    if docker_image_exists(&agent_tag)? && docker_image_exists(&rest_api_tag)? {
+        return Ok(());
+    }
+
+    let tempdir = tempdir()?;
+
+    cmd_exec_checked(
+        Command::new("git")
+            .args([
+                "clone",
+                "--single-branch",
+                "--branch",
+                &format!("v{SDK_VERSION}"),
+                "https://github.com/risc0/risc0.git",
+            ])
+            .arg(tempdir.path()),
+    )?;
+
+    cmd_exec_checked(
+        Command::new("docker")
+            .arg("compose")
+            .arg("--file")
+            .arg(tempdir.path().join("compose.yml"))
+            .arg("--env-file")
+            .arg(tempdir.path().join("bento/dockerfiles/sample.env"))
+            .arg("build"),
+    )?;
+
+    docker_image_tag("agent", agent_tag)?;
+    docker_image_tag("bento-rest_api", rest_api_tag)?;
+
+    Ok(())
+}
+
+const BENTO_COMPOSE_TEMPLATE: &str = include_str!("../../compose.yml");
+const BENTO_GPU_PROVER_AGENT_TEMPLATE: &str = include_str!("../../gpu_prover_agent.yml");
+
+fn bento_compose_file() -> String {
+    let cuda_visible_devices = env::var("CUDA_VISIBLE_DEVICES").unwrap_or_else(|_| "".to_string());
+    let cuda_visible_devices = cuda_visible_devices
+        .split(",")
+        .flat_map(|device_id| device_id.parse::<usize>().ok())
+        .collect::<Vec<_>>();
+
+    let mut compose: serde_yaml::Mapping = serde_yaml::from_str(BENTO_COMPOSE_TEMPLATE).unwrap();
+    let gpu_prover_agent: serde_yaml::Mapping =
+        serde_yaml::from_str(BENTO_GPU_PROVER_AGENT_TEMPLATE).unwrap();
+
+    if cuda_visible_devices.is_empty() {
+        let mut gpu_prover_agent = gpu_prover_agent.clone();
+        gpu_prover_agent.remove("deploy").unwrap();
+        compose["services"]
+            .as_mapping_mut()
+            .unwrap()
+            .insert("gpu_prove_agent0".into(), gpu_prover_agent.into());
+    } else {
+        for idx in cuda_visible_devices {
+            let mut gpu_prover_agent = gpu_prover_agent.clone();
+            gpu_prover_agent["deploy"]["resources"]["reservations"]["devices"][0]["device_ids"]
+                [0] = idx.to_string().into();
+            compose["services"].as_mapping_mut().unwrap().insert(
+                format!("gpu_prove_agent{idx}").into(),
+                gpu_prover_agent.into(),
+            );
+        }
+    }
+
+    serde_yaml::to_string(&compose).unwrap()
+}
+
+/// Execute `docker compose ... {command}` with `bento_compose_file()`.
+fn docker_compose_bento<I, S>(command: I) -> Result<(), io::Error>
+where
+    I: Clone + IntoIterator<Item = S>,
+    S: AsRef<OsStr>,
+{
+    let rust_log = env::var("RUST_LOG").unwrap_or_else(|_| "info".to_string());
+    let segment_size = env::var("SEGMENT_SIZE").unwrap_or_else(|_| "21".to_string());
+    let risc0_keccak_po2 = env::var("RISC0_KECCAK_PO2").unwrap_or_else(|_| "17".to_string());
```

Comment:
Are these defaults coming from the Risc0 repo?

Reply:
Yes, coming from their default env https://github.com/risc0/risc0/blob/v2.3.1/bento/dockerfiles/sample.env. The rest I already subsitute into the yaml file.